### ISSUE TITLE
Congestion control

### ISSUE THREAD

**Issue by avasenin** [TakahikoKawasaki/nv-websocket-client#21]
Do you have any plan to add congestion control to you library?

I was able to use TCP congestion control for the reading thread with current implementation by adding control logic to user's callbacks but have no idea how to add congestion control to writing thread. Right now I get OOM if i'm send more then could transfer to the server because all frames are stored in unlimited linked list before they are sent to server.

**Comment by TakahikoKawasaki** [comment]
Currently, I have no plan to add congenstion control.

How do you expect congention control to work? For example, should `sendXxx` methods block when the number of queued frames is too large?

**Comment by avasenin** [comment]
It's a one of possible option (probably best one). Another possible option is to throw exception. In any case it will be great to have ability to control congestion somehow because right now it's just crash application by OOM.

I found a way how I could implement congestion control for writing thread on top of your library by putting frame to LinkedBlockingQueue before send to websocket and listening onFrameSent to control send flow.

If you don't want to implement congestion control out-of-box I assume that you could close this ticket.

**Comment by avasenin** [comment]
I've tried to implement suggested approach but code looks like a mess.

It will be great to have something useful for congestion from library. It could be
* BlockingLinkedQueue mFrames in WritingThread with ability to specify the length of queue somewhere (factory?). sendFrames could use this queue as is without creating new ArrayList buffer for current frames. Moreover, I suggest that this simplify waiting logic in WritingThread.
* Method to send frames synchronously.
* New callback before/after sendFrame(frame) is called. In this case users could control the flow in this callback. Right now I need to listen all onFrameSent onFrameUnsent and onSendError to understand when frame is passed to Socket.

**Comment by TakahikoKawasaki** [comment]
Released version `1.15`. Please try `WebSocket.setFrameQueueSize(int)` method. The following is an excerpt from `README.md`.

<hr/>
### Congestion Control

`sendXxx` methods queue a `WebSocketFrame` instance to the internal queue. By default, no upper limit is imposed on the queue size, so `sendXxx` methods do not block. However, this behavior may cause a problem if your WebSocket client application sends too many WebSocket frames in a short
time for the WebSocket server to process. In such a case, you may want `sendXxx` methods to block when many frames are queued.

You can set an upper limit on the internal queue by calling `setFrameQueueSize(int)` method. As a result, if the number of frames in the queue has reached the upper limit when a `sendXxx` method is called, the method blocks until the queue gets spaces. The code snippet below is an example to set 5 as the upper limit of the internal frame queue.

```java
// Set 5 as the frame queue size.
ws.setFrameQueueSize(5);
```

Note that under some conditions, even if the queue is full, `sendXxx` methods do not block. For example, in the case where the thread to send frames (`WritingThread`) is going to stop or has already stopped. In addition, method calls to send a [control frame](https://tools.ietf.org/html/rfc6455#section-5.5) (e.g. `sendClose()` and `sendPing()`) do not block.

<hr/>

In addition, a new callback method, `onSendingFrame(WebSocket, WebSocketFrame)` has been added to `WebSocketListener`. It is called right before a frame is sent to the server.

**Comment by avasenin** [comment]
Cool! My two cents:

1) Your implementation cause deadlock and leak of resources when writing thread is waiting for space in mFrames and disconnect happens. It happens because of mStateManager locking isn't not released when you wait on writing thread.
 
 ```
ReadingThread <--- Frozen for at least 3m 49 sec
com.neovisionaries.ws.client.WebSocket.sendFrame(WebSocketFrame) WebSocket.java:1904
com.neovisionaries.ws.client.ReadingThread.readFrame() ReadingThread.java:342
com.neovisionaries.ws.client.ReadingThread.main() ReadingThread.java:93
com.neovisionaries.ws.client.ReadingThread.run() ReadingThread.java:61



WritingThread <--- Frozen for at least 3m 49 sec
com.neovisionaries.ws.client.WebSocket.disconnect(int, String) WebSocket.java:1785
com.neovisionaries.ws.client.WebSocket.disconnect() WebSocket.java:1699
com.machinezone.platform.transport.WebSocketTransport.close() WebSocketTransport.java:46
com.machinezone.platform.connection.Connection.dispose() Connection.java:202
com.machinezone.platform.ClientImpl.disposeConnection() ClientImpl.java:162
com.machinezone.platform.ClientImpl.onConnected(StateMode) ClientImpl.java:128
com.machinezone.platform.fsm.ConnectionFSM$ConnectedState.leave(ConnectionFSM) ConnectionFSM.java:233
com.machinezone.platform.fsm.ConnectionFSM$ConnectedState.leave(AbstractFSM) ConnectionFSM.java:223
com.machinezone.platform.fsm.State.leaveSafe(AbstractFSM) State.java:31
com.machinezone.platform.fsm.State.transition(AbstractFSM, State) State.java:41
com.machinezone.platform.fsm.ConnectionFSM$ConnectedState.onInternalError(ConnectionFSM) ConnectionFSM.java:248
com.machinezone.platform.fsm.ConnectionFSM$6.run() ConnectionFSM.java:87
java.util.concurrent.Executors$RunnableAdapter.call() Executors.java:439
java.util.concurrent.FutureTask$Sync.innerRun() FutureTask.java:303
java.util.concurrent.FutureTask.run() FutureTask.java:138
com.machinezone.platform.utils.TrampolineExecutorService.execute(Runnable) TrampolineExecutorService.java:49
java.util.concurrent.AbstractExecutorService.submit(Runnable) AbstractExecutorService.java:78
com.google.common.util.concurrent.AbstractListeningExecutorService.submit(Runnable) AbstractListeningExecutorService.java:50
com.google.common.util.concurrent.AbstractListeningExecutorService.submit(Runnable) AbstractListeningExecutorService.java:37
com.machinezone.platform.fsm.ConnectionFSM.onInternalError() ConnectionFSM.java:84
com.machinezone.platform.ClientImpl$1.onTransportError(Exception) ClientImpl.java:227
sun.reflect.NativeMethodAccessorImpl.invoke0(Method, Object, Object[]) NativeMethodAccessorImpl.java (native)
sun.reflect.NativeMethodAccessorImpl.invoke(Object, Object[]) NativeMethodAccessorImpl.java:39
sun.reflect.DelegatingMethodAccessorImpl.invoke(Object, Object[]) DelegatingMethodAccessorImpl.java:25
java.lang.reflect.Method.invoke(Object, Object[]) Method.java:597
com.machinezone.platform.connection.Connection$4.invoke(Object, Method, Object[]) Connection.java:311
com.sun.proxy.$Proxy9.onTransportError(Exception)
com.machinezone.platform.transport.WebSocketTransport$1.onError(WebSocket, WebSocketException) WebSocketTransport.java:95
com.neovisionaries.ws.client.ListenerManager.callOnError(WebSocketException) ListenerManager.java:432
com.neovisionaries.ws.client.WritingThread.sendFrame(WebSocketFrame) WritingThread.java:442
com.neovisionaries.ws.client.WritingThread.sendFrames(boolean) WritingThread.java:323
com.neovisionaries.ws.client.WritingThread.main() WritingThread.java:107
com.neovisionaries.ws.client.WritingThread.run() WritingThread.java:57
```
2) The WritingThread frees all elements in once. For example if I set queue length = 100 and queueFrame is locked then I will wait until all previous 100 frames will be sent and new frames will add to intermediate LinkedList which cleans mFrames. It will give me 100 elements quota. I'm expecting then when I sent one frame then I could be able to add one more frame to queue.
 https://github.com/TakahikoKawasaki/nv-websocket-client/blob/master/src/main/java/com/neovisionaries/ws/client/WritingThread.java#L308

3) New callback is awesome. In spite of the issue with deadlock I could provide my own congestion control algorithm and get working application. It's pretty sweet.

**Comment by TakahikoKawasaki** [comment]
Thank you for your feedback. I'll look into the deadlock issue and try to find a solution.

**Comment by TakahikoKawasaki** [comment]
@avasenin
I changed the code for 1) and 2). Could you test `nv-websocket-client-1.16-SNAPSHOT`?
If you cannot test SNAPSHOT, it would be doable to release `1.16` without your testing.

**Comment by TakahikoKawasaki** [comment]
@avasenin I released a new version `1.16`. Please try the new version.

**Comment by avasenin** [comment]
Looks good. I will perform some tests until the end of week. I reopen this issue if I find something.

Thanks for your fast responses and time.
